File: src/main/kotlin/fr/jhelp/kotlinLightSamples/Queue.kt
package fr.jhelp.kotlinLightSamples

import fr.jhelp.kotlinLight.guard
import fr.jhelp.kotlinLigthTools.exceptions.IllegalStateException

class Queue<T>
{
    private var head: Element<T>?
    private var tail: Element<T>?

    constructor()
    {
        this.head = null
        this.tail = null
    }

    fun isEmpty(): Boolean
    {
        return this.head == null
    }

    fun enqueue(element: T)
    {
        if (this.head == null)
        {
            this.head = Element(element)
            this.tail = this.head
            return
        }

        this.tail!!.setNext(element)
        this.tail = this.tail!!.getNext()
    }

    @Throws
    fun dequeue(): T
    {
        (this.head != null).guard { throw IllegalStateException("Queue is empty") }

        val element = this.head!!.getElement()
        this.head = this.head!!.getNext()

        if (this.head == null)
        {
            this.tail = null
        }

        return element
    }
}

File: src/main/swift/Sources/swift/fr/jhelp/kotlinLightSamples/Queue.swift


class Queue<T> {

    private var head: Element<T>?
    private var tail: Element<T>?

    init(){
        self.head = nil
        self.tail = nil
    }

    func isEmpty() -> Bool
    {
        return self.head == nil
    }

    func enqueue(_ element: T)
    {
        if self.head == nil
        {
            self.head = Element(element)
            self.tail = self.head
            return
        }

        self.tail!.setNext(element)
        self.tail = self.tail!.getNext()
    }

    func dequeue() throws  -> T
    {
        guard (self.head != nil) else  { throw CommonManagedExceptions.IllegalStateException("Queue is empty") }

        let element = self.head!.getElement()
        self.head = self.head!.getNext()

        if self.head == nil
        {
            self.tail = nil
        }

        return element
    }
}
